File: packages/java/ygo74-agent-runtime/src/main/java/com/ygo74/agentruntime/routing/RouteRegistry.java
package com.ygo74.agentruntime.routing;

import java.util.HashMap;
import java.util.Map;

public final class RouteRegistry {
    private final Map<String, Dispatcher.UseCaseHandler> handlers = new HashMap<>();

    public void register(String routeKey, Dispatcher.UseCaseHandler handler) {
        if (handlers.containsKey(routeKey)) {
            throw new IllegalArgumentException("route already registered");
        }
        handlers.put(routeKey, handler);
    }

    public Dispatcher.UseCaseHandler resolve(String routeKey) {
        var handler = handlers.get(routeKey);
        if (handler == null) {
            throw new IllegalArgumentException(routeKey);
        }
        return handler;
    }
}

File: packages/dotnet/Ygo74.AgentRuntime/Routing/RouteRegistry.cs
using Ygo74.AgentRuntime.Domains.Contracts;

namespace Ygo74.AgentRuntime.Routing;

public sealed class RouteRegistry
{
    private readonly Dictionary<string, UseCaseHandler> _handlers = new();

    public void Register(string routeKey, UseCaseHandler handler)
    {
        if (_handlers.ContainsKey(routeKey))
        {
            throw new ArgumentException("route already registered", nameof(routeKey));
        }

        _handlers[routeKey] = handler;
    }

    public UseCaseHandler Resolve(string routeKey)
    {
        if (!_handlers.TryGetValue(routeKey, out var handler))
        {
            throw new KeyNotFoundException(routeKey);
        }

        return handler;
    }
}
